common/flatpak-run.c

```diff
@@ -1337,6 +1337,57 @@ flatpak_run_apply_env_appid (FlatpakBwrap *bwrap,
     flatpak_bwrap_set_env (bwrap, "HOST_XDG_CACHE_HOME", g_getenv ("XDG_CACHE_HOME"), TRUE);
 }
 
+static gboolean
+variable_is_safe_to_expand (const char *name)
+{
+  const char *safe_vars[] = {
+    "XDG_DATA_HOME",
+    "XDG_CONFIG_HOME",
+    "XDG_CACHE_HOME",
+    "PATH",
+    "HOME",
+    NULL
+  };
+  return g_strv_contains (safe_vars, name);
+}
+
+static char *
+expand_env_vars (const char *string)
+{
+  g_autoptr(GString) s = g_string_sized_new (256);
+  const char *p = string;
+
+  g_debug ("Expanding string %s", string);
+  while (*p)
+    {
+      if (p[0] == '$' && g_ascii_isalpha (p[1]))
+        {
+          p++;
+          const char *q = p;
+          g_autofree char *name = NULL;
+          while (*p == '_' || g_ascii_isalnum (*p))
+            p++;
+          name = g_strndup (q, p - q);
+          g_debug ("Found variable reference: $%s", name);
+          if (variable_is_safe_to_expand (name))
+            {
+              const char *value = g_getenv (name);
```

Comment:
This isn't right. It should expand it based on the value of the env in bwrap->envp.